File: android/DFM/src/main/java/com/dontflymoney/view/SummaryActivity.java
package com.dontflymoney.view;

import android.app.DatePickerDialog;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.DatePicker;
import android.widget.TableLayout;
import android.widget.TableRow;

import com.dontflymoney.api.Request;
import com.dontflymoney.api.Step;
import com.dontflymoney.baseactivity.SmartActivity;
import com.dontflymoney.viewhelper.TableRowWithExtra;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.lang.reflect.Field;
import java.util.Calendar;

public class SummaryActivity extends SmartActivity
{
	static JSONArray monthList;
	static String name;
	static double total;
	
	TableLayout main;
	String accounturl;

	DatePickerDialog dialog;
	private static int year;



    protected int contentView() { return R.layout.summary; }
    protected int optionsMenuResource() { return R.menu.summary; }



    @Override
	protected void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);
		setCurrentInfo();
		
		if (rotated && succeded)
		{
			setDateFromLast();

			try
			{
				fillSummary();
			}
			catch (JSONException e)
			{
				message.alertError(R.string.error_activity_json, e);
			}
		}
		else
		{
			setDateFromCaller();
			getSummary();
		}
	}
	
	private void setCurrentInfo()
	{
		main = (TableLayout)findViewById(R.id.main_table);
		accounturl = getIntent().getStringExtra("accounturl");
	}
	
	private void setDateFromLast()
	{
		setDate(year);
	}
	
	private void setDateFromCaller()
	{
		Calendar today = Calendar.getInstance();

		int startYear = getIntent().getIntExtra("year", today.get(Calendar.YEAR));
		setDate(startYear);
	}

	private void setDate(int year)
	{
		SummaryActivity.year = year;
		form.setValue(R.id.reportDate, Integer.toString(year));
	}
	
    public void changeDate(View v)
    {
		if(dialog == null)
		{
			dialog = new DatePickerDialog(this, new PickDate(), year, 1, 1);

			try
			{
		        Field pickerField = dialog.getClass().getDeclaredField("mDatePicker");
		        pickerField.setAccessible(true);
		        DatePicker datePicker = (DatePicker) pickerField.get(dialog);
	            
	            Field dayField = datePicker.getClass().getDeclaredField("mDaySpinner");
	            dayField.setAccessible(true);
	            Object dayPicker = dayField.get(datePicker);
	            ((View) dayPicker).setVisibility(View.GONE);
	            
	            Field monthField = datePicker.getClass().getDeclaredField("mMonthSpinner");
	            monthField.setAccessible(true);
	            Object monthPicker = monthField.get(datePicker);
	            ((View) monthPicker).setVisibility(View.GONE);
		    } 
		    catch (Exception e) { }
		}
		
		dialog.show();
    }
	
	private class PickDate implements DatePickerDialog.OnDateSetListener
	{
	    @Override
	    public void onDateSet(DatePicker view, int year, int month, int day)
	    {
            if (view.isShown())
            {
                setDate(year);
                getSummary();
                dialog.dismiss();
            }
	    }
	    
	}
	
	private void getSummary()
	{
		String accounturl = getIntent().getStringExtra("accounturl");
		
		main.removeAllViews();
		
		request = new Request(this, "Moves/Summary");
		
		request.AddParameter("ticket", Authentication.Get());
		request.AddParameter("accounturl", accounturl);
		request.AddParameter("id", year);
		
		request.Post();
	}
	
	@Override
	protected void HandleSuccess(JSONObject data, Step step) throws JSONException
	{
		monthList = data.getJSONArray("MonthList");
		name = data.getString("Name");
		total = data.getDouble("Total");
		
		fillSummary();
	}
	
	private void fillSummary()
		throws JSONException
	{
		form.setValue(R.id.totalTitle, name);
		form.setValueColored(R.id.totalValue, total);
		
		if (monthList.length() == 0)
		{
			View empty = form.createText(getString(R.string.no_summary), Gravity.CENTER);
			main.addView(empty);
		}
		else
		{
			for(int a = 0; a < monthList.length(); a++)
			{
				int color = a % 2 == 0 ? Color.TRANSPARENT : Color.LTGRAY;
				
				getMonth(monthList.getJSONObject(a), color);
			}
		}
	}

	private void getMonth(JSONObject move, int color)
			throws JSONException
	{
		int number = move.getInt("Number");
		
		TableRowWithExtra<Integer> row = new TableRowWithExtra<Integer>(this, number);
		row.setBackgroundColor(color);
		
		String description = move.getString("Name");
		row.addView(form.createText(description, Gravity.LEFT));

		double total = move.getDouble("Total");
		row.addView(form.createText(total, Gravity.RIGHT));

		setClick(row);

		main.addView(row);
	}
	
	private void setClick(TableRow row)
	{
		row.setClickable(true);
		
		row.setOnClickListener(new OnClickListener()
		{
		    public void onClick(View row)
		    {
		    	@SuppressWarnings("unchecked")
				TableRowWithExtra<Integer> tablerow = (TableRowWithExtra<Integer>)row;
		        int monthNumber = tablerow.getExtra();

				Intent intent = new Intent(SummaryActivity.this, ExtractActivity.class);
				
		        intent.putExtra("accounturl", accounturl);
				intent.putExtra("year", year);
				intent.putExtra("month", monthNumber-1);
				
				startActivity(intent);
		    }
		});
	}
	
	

}

File: android/DFM/src/main/java/com/dontflymoney/view/SummaryActivity.kt
package com.dontflymoney.view

import android.app.DatePickerDialog
import android.os.Bundle
import android.view.View
import android.widget.DatePicker
import android.widget.ListView
import android.widget.TextView

import com.dontflymoney.adapters.YearAdapter
import com.dontflymoney.api.InternalRequest
import com.dontflymoney.api.Step
import com.dontflymoney.baseactivity.SmartActivity
import com.dontflymoney.listeners.IDatePickerActivity
import com.dontflymoney.listeners.PickDate

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

import java.lang.reflect.Field
import java.util.Calendar

class SummaryActivity : SmartActivity(), IDatePickerActivity {

    internal var main: ListView
    internal var empty: TextView

    internal var accountUrl: String

    internal var dialog: DatePickerDialog? = null


    override fun contentView(): Int {
        return R.layout.summary
    }

    override fun optionsMenuResource(): Int {
        return R.menu.summary
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setCurrentInfo()

        if (rotated && SmartActivity.succeded) {
            setDateFromLast()

            try {
                fillSummary()
            } catch (e: JSONException) {
                message.alertError(R.string.error_activity_json, e)
            }

        } else {
            setDateFromCaller()
            getSummary()
        }
    }

    private fun setCurrentInfo() {
        main = findViewById(R.id.main_table) as ListView
        empty = findViewById(R.id.empty_list) as TextView

        accountUrl = intent.getStringExtra("accountUrl")
    }

    private fun setDateFromLast() {
        setDate(year)
    }

    private fun setDateFromCaller() {
        val today = Calendar.getInstance()

        val startYear = intent.getIntExtra("year", today.get(Calendar.YEAR))
        setDate(startYear)
    }

    private fun setDate(year: Int) {
        SummaryActivity.year = year
        form.setValue(R.id.reportDate, Integer.toString(year))
    }

    fun changeDate(v: View) {
        if (dialog == null) {
            dialog = DatePickerDialog(this, PickDate(this), year, 1, 1)

            try {
                val pickerField = dialog!!.javaClass.getDeclaredField("mDatePicker")
                pickerField.setAccessible(true)

                val dayField = pickerField.get(dialog).javaClass.getDeclaredField("mDaySpinner")
                dayField.setAccessible(true)
                val dayPicker = dayField.get(pickerField.get(dialog))
                (dayPicker as View).visibility = View.GONE

                val monthField = pickerField.get(dialog).javaClass.getDeclaredField("mMonthSpinner")
                monthField.setAccessible(true)
                val monthPicker = monthField.get(pickerField.get(dialog))
                (monthPicker as View).visibility = View.GONE
            } catch (ignored: Exception) {
            }

        }

        dialog!!.show()
    }

    override fun setResult(year: Int, month: Int, day: Int) {
        setDate(year)
        getSummary()
    }

    private fun getSummary() {
        val accountUrl = intent.getStringExtra("accountUrl")

        request = InternalRequest(this, "Moves/Summary")

        request.AddParameter("ticket", Authentication.Get())
        request.AddParameter("accountUrl", accountUrl)
        request.AddParameter("id", year)

        request.Post()
    }

    override fun getDialog(): DatePickerDialog {
        return dialog
    }

    @Throws(JSONException::class)
    override fun HandleSuccess(data: JSONObject, step: Step) {
        monthList = data.getJSONArray("MonthList")
        name = data.getString("Name")
        total = data.getDouble("Total")

        fillSummary()
    }

    @Throws(JSONException::class)
    private fun fillSummary() {
        form.setValue(R.id.totalTitle, name)
        form.setValueColored(R.id.totalValue, total)

        if (monthList.length() == 0) {
            main.visibility = View.GONE
            empty.visibility = View.VISIBLE
        } else {
            main.visibility = View.VISIBLE
            empty.visibility = View.GONE

            val yearAdapter = YearAdapter(this, monthList, accountUrl, year)
            main.adapter = yearAdapter
        }

    }

    companion object {
        internal var monthList: JSONArray
        internal var name: String
        internal var total: Double = 0.toDouble()
        private var year: Int = 0
    }


}
